packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -830,6 +832,186 @@ function commitLifeCycles(
   );
 }
 
+function toggleHideAndEffectsForSubtree(finishedWork, effectsToRun, isUnmount) {
+  return toggleHideAndEffectsForSubtreeImpl(
+    finishedWork,
+    finishedWork,
+    effectsToRun,
+    isUnmount,
+    true,
+  );
+}
+
+function toggleHideAndEffectsForSubtreeImpl(
+  finishedWork,
+  fiber,
+  effectsToRun,
+  isUnmount,
+  shouldToggleHide,
+) {
+  if (fiber === null) {
+    return;
+  } else if (
+    (fiber.tag === OffscreenComponent || fiber.tag === LegacyHiddenComponent) &&
+    finishedWork !== fiber &&
+    fiber.memoizedState !== null
+  ) {
+    // A nested Offscreen component is Offscreen, which means its effects have already
+    // been unmounted. Skip over the children.
+    toggleHideAndEffectsForSubtreeImpl(
+      finishedWork,
+      fiber.sibling,
+      effectsToRun,
+      isUnmount,
+      shouldToggleHide,
+    );
+  } else {
+    let shouldToggleHideForChildren = shouldToggleHide;
+    switch (fiber.tag) {
+      case HostComponent: {
+        if (shouldToggleHide) {
+          const instance = fiber.stateNode;
+          if (isUnmount) {
+            hideInstance(instance);
+          } else {
+            unhideInstance(fiber.stateNode, fiber.memoizedProps);
+          }
+          // Subsequent children do not need their hidden/unhidden because the
+          // parent has been hidden/unhidden
+          shouldToggleHideForChildren = false;
+        }
+
+        if (fiber.ref !== null) {
+          if (isUnmount) {
+            commitDetachRef(fiber);
+          } else {
+            fiber.effectTag |= Ref;
+          }
+        }
+        break;
+      }
+      case HostText: {
+        if (shouldToggleHide) {
+          const instance = fiber.stateNode;
+          if (isUnmount) {
+            hideTextInstance(instance);
+          } else {
+            unhideTextInstance(instance, fiber.memoizedProps);
+          }
+          // Subsequent children do not need to be hidden/unhidden because the
+          // parent has been hidden/unhidden
+          shouldToggleHideForChildren = false;
+        }
+        break;
+      }
+      case ClassComponent: {
+        if (isUnmount) {
+          if (fiber.ref !== null) {
+            commitDetachRef(fiber);
+          }
+          const instance = fiber.stateNode;
+          if (typeof instance.componentWillUnmount === 'function') {
+            safelyCallComponentWillUnmount(fiber, instance);
+          }
+        } else {
+          if (fiber.ref !== null) {
+            fiber.effectTag |= Ref;
+          }
+
+          const instance = fiber.stateNode;
+          if (typeof instance.componentDidMount === 'function') {
+            fiber.effectTag |= Update | Placement;
```

Comment:
Hmm. This isn't going to work, because it assumes that the class fiber is already part of the effect list. That's not always the case. In the begin phase, we might have bailed out before reaching that node. Like if a parent were wrapped in `memo`.

Same goes for hooks effect below.

Need to think about the right solution. Ideally you would call the effect immediately within this function, but it's too early. We're in the mutation phase, not the layout phase.

We could push the effects into an array so we can fire them in the next phase. Like we do for passive effects.